hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyServiceImpl.java

```diff
@@ -243,7 +245,12 @@ public void dispatchEvent(final DistributedObjectEventPacket eventPacket, Object
         if (eventPacket.getEventType() == CREATED) {
             try {
                 final ProxyRegistry registry = getOrCreateRegistry(serviceName);
-                if (!registry.contains(eventPacket.getName())) {
+                if (registry.contains(eventPacket.getName())) {
+                    // proxy exists, make sure it is initialized
+                    // listeners will not be called if the proxy is initialized
+                    // here (they were called when the proxy was created)
+                    registry.initializeProxy(eventPacket.getName(), false);
```

Comment:
Generally I'm ok with the solution. The only concern I have is that this is a call which potentially invokes operations on other members. In this case, we're blocking the [event service thread](https://github.com/hazelcast/hazelcast/blob/c29e1d945fb13af172420b058093df1bcf3ec304/hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java#L183-L187).

Now, maybe I just have false concerns but because of a [recent issue with deadlocks in proxy creation](https://github.com/hazelcast/hazelcast/pull/15740), I'm thinking if the same can happen here. WDYT?

Reply:
No, this call is local-only (as a result of receiving an event from a remote member). The `false` argument to `initializeProxy()` makes sure that no events are published to other members. So I think we are safe here.

Reply:
No, this call is local-only (as a result of receiving an event from a remote member). The `false` argument to `initializeProxy()` makes sure no events are published to other members. So I think we are safe here.

Reply:
But, still, the `future.get();` is waiting on some `InitializingObject#initialize` method to complete which in turn may invoke operations on other members, which may cause an event to be fired back at the caller. Or not? I'm sure about the first part of invoking operations on other members, the question is if there is some piece of code somewhere that would complete the circle.

Reply:
Ah, you are right - `InitializingObject#initialize` may trigger creation of indexes on other members. So you can have a situation that index creation on member A triggers index creation on member B which in turn triggers index creation on A. Need to think about this more.

Reply:
I can't put my finger on it, maybe we're safe, maybe we're not. We just need to think about it a bit.

Reply:
@mmedenjak, I believe I have found a simple way to resolve the concurrency issue in a recent commit https://github.com/hazelcast/hazelcast-enterprise/pull/3343/commits/cbc97ae77ee5d4696794ff7c550197622ed20030 in https://github.com/hazelcast/hazelcast-enterprise/pull/3343 based on the original 'force-init' PR https://github.com/hazelcast/hazelcast/pull/15986. If it proves to work, then this PR is no longer necessary and can be closed.

Reply:
Great, will take a look.